lib/libvfio-user.c

```diff
@@ -1052,6 +1052,45 @@ handle_dirty_pages(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
     return ret;
 }
 
+static int
+handle_device_feature(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+{
+    struct vfio_user_device_feature *req = msg->in.iov.iov_base;
+
+    if (vfu_ctx->migration == NULL) {
+        return -EINVAL;
+    }
+
+    if (!migration_feature_supported(req->flags)) {
+        // FIXME what error code to return? we really want "not supported"
+        //   instead of "not permitted"?
+        return -EINVAL;
+    }
+
+    ssize_t ret;
+
+    if (req->flags & VFIO_DEVICE_FEATURE_PROBE) {
+        msg->out.iov.iov_base = msg->in.iov.iov_base;
```

Comment:
This implementation doesn't make much sense to me: it has no effect?

Reply:
We set the reply to equal the request, as this is what VFIO and hence vfio-user dictates for the success response - this has since been replaced by the following, as the aliasing caused a double free: 😳

```c
msg->out.iov.iov_base = malloc(msg->in.iov.iov_len);
msg->out.iov.iov_len = msg->in.iov.iov_len;
memcpy(msg->out.iov.iov_base, msg->in.iov.iov_base,
       msg->out.iov.iov_len);
```

Reply:
Such an API would not be useful. I think you're supposed to respond with the set of flags known to the server. (And, if get, or set, is set, whether the server supports that verb for the given flag(s)).

Reply:
Are we sure we want to diverge from VFIO here?

From documentation of VFIO_DEVICE_FEATURE:
> Get, set, or probe feature data of the device.  The feature is selected using the FEATURE_MASK portion of the flags field.  Support for a feature can be probed by setting both the FEATURE_MASK and PROBE bits.  A probe may optionally include the GET and/or SET bits to determine read vs write access of the feature respectively.  Probing a feature will return success if the feature is supported and all of the optionally indicated GET/SET methods are supported.  The format of the data portion of the structure is specific to the given feature.  The data portion is not required for probing.  GET and SET are mutually exclusive, except for use with PROBE.